### PR TITLE
Orca 679 global atlantis lock new release branch

### PR DESCRIPTION
**Pull request by msarvar**
Added support to create a command specific lock entries to the db/bolt_db, LockCommand, UnlockCommand, and CheckCommandLock.

Implemented a new ApplyLocker interface for locking.go#Client that manages apply lock creation/deletion and lock retrieval.

LocksController has 2 handlers to create and delete apply locks.
Server#Index renders the UI to create/delete the locks.
ApplyCommandRunner expects an object that implements ApplyCommandLocker which boils down to object having IsDisabled function that returns boolean value.
DefaultApplyCommandLocker implements the interface.

In UI it is just a AJAX request that will create/delete the locks for apply. To create a new lock or delete existing one do to atlantis main page.

Recreated #48 against new release branch.

### CODE REVIEW

File: server/events/apply_command_runner.go [hunk 1]
```diff
@@ -2,14 +2,49 @@ package events
...
 	"github.com/runatlantis/atlantis/server/events/db"
+	"github.com/runatlantis/atlantis/server/events/locking"
 	"github.com/runatlantis/atlantis/server/events/models"
 	"github.com/runatlantis/atlantis/server/events/vcs"
 )
 
+//go:generate pegomock generate -m --use-experimental-model-gen --package mocks -o mocks/mock_apply_command_locker.go ApplyCommandLocker
+
+type ApplyCommandLocker interface {
+	IsDisabled(ctx *CommandContext) bool
```
Comment (nishkrishnan) [inline server/events/apply_command_runner.go:13]: ie. This struct isn't responsible for performing the locking, it's responsible for checking whether a lock exists. Confusing naming imo.

Id suggest renaming this to `GlobalApplyLockChecker` with the method `isLocked` (disabled also doesn't fit well if the name of the struct doesn't include that)

File: server/events/apply_command_runner.go [hunk 2]
```diff
@@ -2,14 +2,49 @@ package events
...
+}
+
+func NewApplyCommandLocker(
+	applyLockChecker locking.ApplyLockChecker,
+	disableApply bool,
```
Comment (nishkrishnan) [inline server/events/apply_command_runner.go:18]: can we rename this to disableApplyFlag or something that hints at this being a global server flag?

File: server/events/apply_command_runner.go [hunk 3]
```diff
@@ -2,14 +2,49 @@ package events
...
+
+type DefaultApplyCommandLocker struct {
+	ApplyLockChecker locking.ApplyLockChecker
+	DisableApply     bool
+}
+
+// IsDisabled returns true if there is a global apply command lock or
+// DisableApply flag is set to true
+func (a *DefaultApplyCommandLocker) IsDisabled(ctx *CommandContext) bool {
+	lock, err := a.ApplyLockChecker.CheckApplyLock()
```
Comment (nishkrishnan) [inline server/events/apply_command_runner.go:34]: Looks like command context is barely used here. The logger embedded in there is useful if we want to debug per-pull issues. However, the global apply lock fetching error doesn't seem necessary to log at that level, you can just keep that logger embedded in the struct.


If you do the above, you can just use a single interface delegation pattern for:
1) fetching the global lock
2) comparing the result of that to the global server config flag.

File: server/events/apply_command_runner.go [hunk 4]
```diff
@@ -53,7 +88,7 @@ func (a *ApplyCommandRunner) Run(ctx *CommandContext, cmd *CommentCommand) {
 	baseRepo := ctx.Pull.BaseRepo
 	pull := ctx.Pull
 
-	if a.DisableApply {
+	if a.locker.IsDisabled(ctx) {
 		ctx.Log.Info("ignoring apply command since apply disabled globally")
 		if err := a.vcsClient.CreateComment(baseRepo, pull.Num, applyDisabledComment, models.ApplyCommand.String()); err != nil {
```
Comment (nishkrishnan) [inline server/events/apply_command_runner.go:93]: The applyDisabledComment is very generic, would be good to think about how to make that more useful depending on how the apply command lock is configured.  Not in scope for this PR though.

File: server/events/locking/locking.go [hunk 5]
```diff
@@ -66,6 +98,49 @@ func NewClient(backend Backend) *Client {
...
+	err := c.backend.UnlockCommand(models.ApplyCommand)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckApplyLock retrieves an apply command lock if present.
+func (c *Client) CheckApplyLock() (ApplyCommandLockResponse, error) {
```
Comment (nishkrishnan) [inline server/events/locking/locking.go:128]: Sounds to me like this should also follow the same logic of checking the server configuration as well. We should also probably not allow unlocking of atlantis if it's configured to disable applies through server configuration both.

Right now the implementation doesn't do that which is good, however, it'll say it's locked and unlocked it, but applies still won't work if the server configuration has it defined.

**Review by nishkrishnan — COMMENTED** [review]
Nice work, have a few comments

File: server/events/locking/locking.go [hunk 6]
```diff
@@ -66,6 +98,49 @@ func NewClient(backend Backend) *Client {
 	}
 }
 
+// LockApply acquires global apply lock.
+func (c *Client) LockApply() (ApplyCommandLockResponse, error) {
+	response := ApplyCommandLockResponse{}
+
+	applyCmdLock, err := c.backend.LockCommand(models.ApplyCommand, time.Now().Local())
```
Comment (nishkrishnan) [inline server/events/locking/locking.go:105]: Best practice is to use unix timestamps if we are storing data.

**Review by nishkrishnan — DISMISSED** [review]

File: server/events/command_runner_test.go [hunk 7]
```diff
@@ -86,10 +88,12 @@ func setup(t *testing.T) *vcsmocks.MockClient {
 
 	drainer = &events.Drainer{}
 	deleteLockCommand = eventmocks.NewMockDeleteLockCommand()
+	applyLockChecker = lockingmocks.NewMockApplyLockChecker()
 	When(logger.GetLevel()).ThenReturn(logging.Info)
 	When(logger.NewLogger("runatlantis/atlantis#1", true, logging.Info)).
 		ThenReturn(pullLogger)
 
+	scope := stats.NewStore(stats.NewNullSink(), false)
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner_test.go:96]: No, this is to remove the logging warnings about statsd failing to reach the localhost. Just to reduce the noise :)

**Review by nishkrishnan — APPROVED** [review]

**Comment by msarvar** [comment]
💨

**Pull request merged**